featurebyte/service/feature_table_cache.py

```diff
@@ -0,0 +1,153 @@
+):
+    """
+    Feature Table Cache metadata service
+    """
+
+    document_class = FeatureTableCacheModel
+
+    def __init__(
+        self,
+        user: Any,
+        persistent: Persistent,
+        catalog_id: Optional[ObjectId],
+        block_modification_handler: BlockModificationHandler,
+        observation_table_service: ObservationTableService,
+    ):
+        super().__init__(
+            user=user,
+            persistent=persistent,
+            catalog_id=catalog_id,
+            block_modification_handler=block_modification_handler,
+        )
+        self.observation_table_service = observation_table_service
+
+    async def get_document_for_observation_table(
+        self,
+        observation_table_id: PydanticObjectId,
+    ) -> Optional[FeatureTableCacheModel]:
+        """Get document for observation table.
+
+        Parameters
+        ----------
+        observation_table_id: PydanticObjectId
+            Observation table id
+
+        Returns
+        -------
+        Optional[FeatureTableCacheModel]
+            Feature Table Cache model if exists
+        """
+        documents = []
+
+        query_filter = {"observation_table_id": observation_table_id}
+        async for document in self.list_documents_iterator(query_filter=query_filter):
+            documents.append(document)
+
+        if not documents:
+            return None
+
+        return documents[0]
+
+    async def get_feature_table_cache_info(
+        self,
+        observation_table_id: PydanticObjectId,
+    ) -> Optional[FeatureTableCacheInfo]:
+        """Get document for observation table.
+
+        Parameters
+        ----------
+        observation_table_id: PydanticObjectId
+            Observation table id
+
+        Returns
+        -------
+        Optional[FeatureTableCacheInfo]
+            Feature Table Cache Info
+        """
+        document = await self.get_document_for_observation_table(observation_table_id)
+        if not document:
+            return None
+
+        return FeatureTableCacheInfo(
+            cache_table_name=document.name,
+            cached_feature_names=document.features,
+        )
+
+    async def update_feature_table_cache(
+        self,
+        observation_table_id: PydanticObjectId,
+        feature_definitions: List[CachedFeatureDefinition],
+    ) -> None:
+        """
+        Update Feature Table Cache by adding new feature definitions.
+
+        Parameters
+        ----------
+        observation_table_id: PydanticObjectId
+            Observation table id
+        feature_definitions: List[CachedFeatureDefinition]
+            Feature definitions
+        """
+        features = []
+        for feature_def in feature_definitions:
+            feature_name = feature_def.name
+            if feature_def.definition_hash:
+                feature_name = f"{feature_name}_{feature_def.definition_hash}"
```

Comment:
Features with different names but the same definition hash are considered identical features and can share the same cached column. Is it intended they all get added to the `features` list? How should the `features` field in `FeatureTableCacheModel` be used?

From caller's perspective, one key information this needs to provide is for each feature definition, is the result already cached, and what is the table name and column name of the cached result.

Reply:
We can probably only use definition_hash as a feature name. 

I see the flow in a following way: 
- We have `observation_table_id` and feature `definition_hash`. We want to determine if feature is already cached. 
- We lookup  `FeatureTableCacheModel` by `observation_table_id`, and check if `definition_hash` is present in the `features` field. 
- If it does - feature is cached. 
- If it doesn't - we cache the feature in the `table_name` (name field which I will rename), and add feature to `features` list.

Reply:
Alternative flow can be to have the model flat, e.g. 

```python 
class FeatureTableCacheModel: 
    table_name: str 
    observation_table_id: ObjectId
    feature_definition_hash: str  
```

I think this approach is a bit safer in terms of 16mb limit, but at the same time I think it will be quite hard to create a FeatureTableCacheModel with so many features so they exceed the limit.